File: bwl.Framework/src/AutoUI/RemoteElements/RemoteAutoListbox.vb
Imports System.Drawing
Imports Bwl.Framework

Public Class RemoteAutoListbox
    Inherits BaseRemoteElement

    Public ReadOnly Property AutoHeight As Boolean

    Public Sub New()
        Me.New(New UIElementInfo("", ""))
    End Sub

    Public Sub New(info As UIElementInfo)
        MyBase.New(info)
        InitializeComponent()
        AddHandler info.Changed, AddressOf BaseInfoChanged
        BaseInfoChanged(info)
    End Sub

    Private Sub BaseInfoChanged(source As UIElementInfo)
        If InvokeRequired Then
            Me.Invoke(Sub() BaseInfoChanged(source))
        Else
            ElementCaption.Text = Info.Caption
            If Info.BackColor.A = 255 Then If ListBox1.BackColor <> Info.BackColor Then ListBox1.BackColor = Info.BackColor
            If Info.ForeColor.A = 255 Then If ListBox1.ForeColor <> Info.ForeColor Then ListBox1.ForeColor = Info.ForeColor
            If Info.Width > 0 Then If Me.Width <> Info.Width Then Me.Width = Info.Width
            If Info.Height > 0 Then If Me.Height <> Info.Height Then Me.Height = Info.Height
        End If
    End Sub

    Public Overrides Sub ProcessData(dataname As String, data() As Byte)
        If dataname.ToLower = "items" Then
            Dim items = AutoUIByteCoding.GetParts(data)
            Me.Invoke(Sub()
                          SyncLock ListBox1
                              Dim currItems = ""
                              Dim newItems = ""
                              For Each item As String In Me.ListBox1.Items : currItems += item : Next
                              For Each item As String In items : newItems += item : Next
                              If currItems <> newItems Then
                                  Do While Me.ListBox1.Items.Count > items.Length
                                      ListBox1.Items.RemoveAt(ListBox1.Items.Count - 1)
                                  Loop
                                  Do While Me.ListBox1.Items.Count < items.Length
                                      ListBox1.Items.Add("")
                                  Loop
                                  For i = 0 To items.Length - 1
                                      ListBox1.Items(i) = items(i)
                                  Next

                                  If AutoHeight Then
                                      Dim newHeight = (ListBox1.Items.Count + 1) * ListBox1.ItemHeight + 20
                                      If Me.Height <> newHeight Then Me.Height = newHeight
                                  End If
                              End If
                          End SyncLock
                      End Sub)
        End If
        If dataname.ToLower = "parameters" Then
            Dim items = AutoUIByteCoding.GetParts(data)
            Me.Invoke(Sub()
                          _AutoHeight = (items(0) = "True")
                          If AutoHeight Then
                              Dim newHeight = (ListBox1.Items.Count + 1) * ListBox1.ItemHeight + 20
                              If Me.Height <> newHeight Then Me.Height = newHeight
                          End If
                      End Sub)
        End If
        If dataname.ToLower = "setselected" Then
            Dim items = AutoUIByteCoding.GetParts(data)
            Me.Invoke(Sub()
                          Dim idx = CType(items(0), Integer)
                          If ListBox1.Items.Count > 0 AndAlso idx < ListBox1.Items.Count Then
                              ListBox1.SetSelected(idx, True)
                          End If
                      End Sub)
        End If
    End Sub

    Private Sub ListBox1_Click(sender As Object, e As EventArgs) Handles ListBox1.Click
        Send("click", {ListBox1.SelectedIndex.ToString})
    End Sub

    Private Sub ListBox1_DoubleClick(sender As Object, e As EventArgs) Handles ListBox1.DoubleClick
        Send("double-click", {ListBox1.SelectedIndex.ToString})
    End Sub

    Private Sub ListBox1_SelectedIndexChanged(sender As Object, e As EventArgs) Handles ListBox1.SelectedIndexChanged
        Send("selected-index-changed", {ListBox1.SelectedIndex.ToString})
    End Sub

    Private Sub RemoteAutoListbox_Load(sender As Object, e As EventArgs) Handles MyBase.Load

    End Sub
End Class

File: Framework/Framework/Bwl.Framework.Windows/AutoUI/RemoteElements/RemoteAutoListbox.cs
using System;
using System.Runtime.Versioning;

namespace Bwl.Framework.Windows
{

    [SupportedOSPlatform("windows")]
    public partial class RemoteAutoListbox : BaseRemoteElement
    {

        public bool AutoHeight { get; private set; }

        public RemoteAutoListbox() : this(new UIElementInfo("", ""))
        {
        }

        public RemoteAutoListbox(UIElementInfo info) : base(info)
        {
            InitializeComponent();
            info.Changed += BaseInfoChanged;
            BaseInfoChanged(info);
        }

        private void BaseInfoChanged(UIElementInfo source)
        {
            if (InvokeRequired)
            {
                Invoke(() => BaseInfoChanged(source));
            }
            else
            {
                ElementCaption.Text = Info.Caption;
                if (Info.BackColor.A == 255)
                {
                    if (ListBox1.BackColor != GetColor(Info.BackColor))
                        ListBox1.BackColor = GetColor(Info.BackColor);
                }
                if (Info.ForeColor.A == 255)
                {
                    if (ListBox1.ForeColor != GetColor(Info.ForeColor))
                        ListBox1.ForeColor = GetColor(Info.ForeColor);
                }
                if (Info.Width > 0)
                {
                    if (Width != Info.Width)
                        Width = Info.Width;
                }
                if (Info.Height > 0)
                {
                    if (Height != Info.Height)
                        Height = Info.Height;
                }
            }
        }

        public override void ProcessData(string dataname, byte[] data)
        {
            if (dataname.ToLower() == "items")
            {
                string[] items = AutoUIByteCoding.GetParts(data);

                Invoke(() => { lock (ListBox1) { string currItems = ""; string newItems = ""; foreach (string item in ListBox1.Items) currItems += item; foreach (string item in items) newItems += item; if ((currItems ?? "") != (newItems ?? "")) { while (ListBox1.Items.Count > items.Length) ListBox1.Items.RemoveAt(ListBox1.Items.Count - 1); while (ListBox1.Items.Count < items.Length) ListBox1.Items.Add(""); for (int i = 0, loopTo = items.Length - 1; i <= loopTo; i++) ListBox1.Items[i] = items[i]; if (AutoHeight) { int newHeight = (ListBox1.Items.Count + 1) * ListBox1.ItemHeight + 20; if (Height != newHeight) Height = newHeight; } } } });
            }
            if (dataname.ToLower() == "parameters")
            {
                string[] items = AutoUIByteCoding.GetParts(data);
                Invoke(() =>
                    {
                        AutoHeight = items[0] == "True";
                        if (AutoHeight)
                        {
                            int newHeight = (ListBox1.Items.Count + 1) * ListBox1.ItemHeight + 20;
                            if (Height != newHeight)
                                Height = newHeight;
                        }
                    });
            }
            if (dataname.ToLower() == "setselected")
            {
                string[] items = AutoUIByteCoding.GetParts(data);
                Invoke(() =>
                    {
                        Int32.TryParse(items[0], out int idx);
                        if (ListBox1.Items.Count > 0 && idx < ListBox1.Items.Count)
                        {
                            ListBox1.SetSelected(idx, true);
                        }
                    });
            }
        }

        private void ListBox1_Click(object sender, EventArgs e)
        {
            Send("click", new[] { ListBox1.SelectedIndex.ToString() });
        }

        private void ListBox1_DoubleClick(object sender, EventArgs e)
        {
            Send("double-click", new[] { ListBox1.SelectedIndex.ToString() });
        }

        private void ListBox1_SelectedIndexChanged(object sender, EventArgs e)
        {
            Send("selected-index-changed", new[] { ListBox1.SelectedIndex.ToString() });
        }

        private void RemoteAutoListbox_Load(object sender, EventArgs e)
        {

        }
    }
}